zerver/actions/users.py

```diff
@@ -94,6 +96,93 @@ def do_delete_user(user_profile: UserProfile, *, acting_user: Optional[UserProfi
         )
 
 
+def do_delete_user_preserving_messages(user_profile: UserProfile) -> None:
+    """
+    This is a version of do_delete_user which does not delete message content.
+
+    The code is a bit tricky, because we want to, at some point, call
+    user_profile.delete() to trigger cascade deletes to clean up the variety
+    of related objects - but we need to avoid the cascades affecting
+    messages sent by the user.
+    """
+    if user_profile.realm.is_zephyr_mirror_realm:
+        raise AssertionError("Deleting zephyr mirror users is not supported")
+
+    do_deactivate_user(user_profile, acting_user=None)
+
+    subscribed_huddle_recipient_ids = set(
+        Subscription.objects.filter(
+            user_profile=user_profile, recipient__type=Recipient.HUDDLE
+        ).values_list("recipient_id", flat=True)
+    )
+    user_id = user_profile.id
+    personal_recipient = user_profile.recipient
+    realm = user_profile.realm
+    date_joined = user_profile.date_joined
+
+    with transaction.atomic():
+        # The strategy is that before calling user_profile.delete(), we need to
+        # reassign Messages  sent by the user to a dummy user, so that they don't
+        # get affected by CASCADE. We cannot yet create a dummy user with .id
+        # matching that of the user_profile, so the general scheme is:
+        # 1. We create a *temporary* dummy for the initial re-assignment of messages.
+        # 2. We delete the UserProfile.
+        # 3. We create a replacement dummy user with its id matching what the UserProfile had.
+        # 4. This is the intended, final replacement UserProfile, so we re-assign
+        #    the messages from step (1) to it and delete the temporary dummy.
+        random_token = secrets.token_hex(16)
+        temp_replacement_user = create_user(
```

Comment:
Can `create_user` have side effects like writing to a cache? It does call `.save()` on a UserProfile object.

Maybe it's OK to pollute caches with this delete user ID, because nothing will access it, but seems worth noting explicitly.

Reply:
Hmm, I'm not sure why it would pollute caches? It does `.save()`, but doesn't that only flush caches? (not even that in this case, since it's a new user profile, so there's nothing to flush)

Reply:
Good point -- I think it's possible that I muddled the waters on this at some point.  We do a [read-through on cache miss](https://github.com/zulip/zulip/blob/5a9cba1bba9eeac4564835bf0e827ed8f89060d6/zerver/lib/cache.py#L176), filling the cache, but we don't ever to write-throughs.